walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,64 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.sql.Time;
+import java.util.*;
 
 public final class FindMeetingQuery {
+  /**
+   * Returns the times when {@code request} could happen on a day containing {@code events}
+   * Time complexity of the algorithm is O(nlogn + mn), in which
+   * n is the number of {@code events}, and m is the number of attendees in {@code request}
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    Collection<TimeRange> availableTimes = new ArrayList<>();
+    // Handle invalid request duration and no attendees edge case in request
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return availableTimes;
+    } else if (request.getAttendees().size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Sort events based on event start time
+    List<Event> eventList = new ArrayList<>(events);
+    eventList.sort(Comparator.comparingInt(e -> e.getWhen().start()));
+
+    // Locate gaps between events joined by any of the request attendees
```

Comment:
Everytime you have to write a comment like this to explain the logic, think about whether you can extract the logic as a method with a proper name instead.

Reply:
Is there a need to refactor even further the whole section when it is the main part of the algorithm (from after sorting the list to right before return)?

Reply:
Maybe and maybe not, it depends.
Some general rules that you can apply:
* If you see yourself having to write a comment to explain section of your code, maybe it's a sign that you can refactor it to a method with self-explanatory name
* Refactor logic of the code that requires cognitive power to understand. Hide the details behind a well-named method so that the reader doesn't have to spend cognitive load to understand the section.

In this example here, you could actually make your code to be like below.
The collection shouldn't be provided, but returned from your method.
```java
Collection<TimeRange> availableTimes = locateGaps(eventList, request);